### PR TITLE
Update the toolbar and panel for the new profiler design

### PR DESCRIPTION
**Pull request by javiereguiluz**
This PR adds support for the new Symfony Toolbar (see https://github.com/symfony/symfony/pull/15160) and it uses the same trick validated by @stof for the DoctrineBundle (see https://github.com/doctrine/DoctrineBundle/pull/445).

### CODE REVIEW

**Comment by stof** [comment]
why are you reverting an empty commit ? It looks to me you are missing the point of ``git revert`` :)

**Comment by sstok** [comment]
@javiereguiluz `git commit --amend` also creates a new commit which you can force push ;)

**Comment by stof** [comment]
anyway, I can just squash the empty commits with gh before merging the changes, but this will wait until the new toolbar design is accepted in symfony itself

**Comment by javiereguiluz** [comment]
This pull request is now ready for the final review. It contains the redesign of both the toolbar and the profiler.

**Comment by WouterJ** [comment]
It's called "E-mails" in the menu and "Swiftmailer" in the page title. That's inconsistent imo and I propose to change the title to "E-mails".

**Comment by WouterJ** [comment]
Also, all other panels show a dashed border + a nice message when no info is shown. In the case of emails, there is a table with only the table head visible (as there is no table body). Imo, there should be the same nice dash-bordered box as in other panels.

**Comment by javiereguiluz** [comment]
@WouterJ about the dashed border, when no emails are sent, the following code is executed:

```twig
        {% if not collector.mailers %}
            <div class="empty">
                <p>No e-mail messages were sent.</p>
            </div>
        {% endif %}
```

At least locally is working for me:

no-email-sent

**Comment by javiereguiluz** [comment]
> It's called "E-mails" in the menu and "Swiftmailer" in the page title. That's inconsistent imo and I propose to change the title to "E-mails".

You are right about this inconsistency. Sadly, most panels use a name different than their menu option. I've fixed it in 43f5639 but maybe the best solution would be to not show the panel name in the page title.

**Comment by WouterJ** [comment]
@javiereguiluz as mentioned in the other PR, I forgot I also had to bring a new version of Swiftmailer. So I actually reviewed the old design :)

**Comment by stof** [comment]
> You are right about this inconsistency. Sadly, most panels use a name different than their menu option. I've fixed it in symfony/symfony@43f5639 but maybe the best solution would be to not show the panel name in the page title.

The solution might be to let the panel provide the text displayed in the title

File: Resources/views/Collector/swiftmailer.html.twig [hunk 1]
```diff
@@ -1,97 +1,229 @@
...
-            <tr>
-                <th scope="col">Mailer</th>
-                <th scope="col">Messages</th>
-            </tr>
-        </thead>
-        <tbody>
-            {% for name in collector.mailers %}
+    {% set profiler_markup_version = profiler_markup_version|default(1) %}
+
+    {% if profiler_markup_version == 1 %}
```
Comment (stof) [inline Resources/views/Collector/swiftmailer.html.twig]: Can we avoid duplicating the whole panel between the old and the new design of the profiler, finding a markup looking fine in both designs ? I would be much easier to maintain

Comment (javiereguiluz) [inline Resources/views/Collector/swiftmailer.html.twig]: I've tried it ... but there are so many little differences between both designs that I cannot deduplicate them.

Comment (stof) [inline Resources/views/Collector/swiftmailer.html.twig]: well, this may mean we should make some changes in the new design PR in Symfony to make it easier to share markup. Maintaining support for both designs at the same time will be a must have for all bundles providing a webprofiler panel for at least 3 years if they want to keep support for the 2.7 LTS.

Comment (WouterJ) [inline Resources/views/Collector/swiftmailer.html.twig]: As I was testing the old design yesterday with the new profiler, it actually didn't look that bad (besides the menu item). Is it really required to have a new design, or is this just a redesign to update all core packages with a new design?

Comment (javiereguiluz) [inline Resources/views/Collector/swiftmailer.html.twig]: The problem here is that we're doing two different things at the same time: 1) update the design of the panel; 2) improve the panel contents.

If the old panel's contents are OK, you only have to do minimal changes ... but that's not the case for this panel.

Comment (WouterJ) [inline Resources/views/Collector/swiftmailer.html.twig]: So that means this PR doesn't actually show the impact of the new profiler design for packages. If they just use the same design as with the old design, they only have to do minimal changes.

I think that's what @stof wants to see: A real example of what a bundle has to update with the redesign. Maybe you can show this by updating another bundle with only the required minimal changes (like SonataBlockBundle)?

Comment (javiereguiluz) [inline Resources/views/Collector/swiftmailer.html.twig]: In EasyAdmin I recently added a profiler compatible with both the old and the new design: https://github.com/javiereguiluz/EasyAdminBundle/pull/413/files

The changes are:

  * I use a single SVG icon for the menu item, but I change its color depending on the profiler version. This is an optional change to make it look "perfect" in both versions.
  * In the old profiler I use a table and in the new one I use the `.metrics` element. Again this is optional and I could use the exact same code for the new profiler. I did this because the metrics look cool and the table is boring.
  * The `.tabs` element of the new profiler degrades gracefully, so you can see thath I reuse the exact asme tabs in the old and the new profiler.

In summary, I could change nothing, but I decided to change one icon color and one table to make it look better in the new profiler.

Comment (stof) [inline Resources/views/Collector/swiftmailer.html.twig]: This PR does 2 things at the same time, but it improves the panel content only for people using the new design. This is where the issue is IMO. Content improvements should be applied to everyone (the content of this panel was not improved since a very long time, and the initial design of the panel was not thought to support multiple mailers, which is probably why it is not that good).

The EasyAdminBundle example is a very good example of what I'm aiming for: having only minor differences between versions, just to make it look better in places where the new version has a better way to display it which would not degrade gracefully in the old version.

**Comment by javiereguiluz** [comment]
@stof I think that in b3dd574 I finally did what you were looking for. We now use the exact same code for both the old and the new profiler ... except for some metrics/tables. The trick is to add some lines of CSS for the old profiler to make it look better.

This is how it looks the latest code of this PR inside an old Symfony application:

**Usual scenario: 1 mailer 1 email**

mailer-4

**1 mailer 2 emails**

mailer-3

**2 mailers but only one sends emails**

mailer-2

**Most complex scenario: several mailers and several emails**

mailer-1

**Comment by stof** [comment]
@javiereguiluz can you post a screenshot for a realistic use case for HTML parts of emails, i.e. a big HTML blob rather than a single line ? I was never able to write a real HTML email with such a small markup :smile:

File: Resources/views/Collector/swiftmailer.html.twig [hunk 2]
```diff
@@ -1,77 +1,175 @@
...
+                <span class="sf-toolbar-status">{{ collector.messageCount }}</span>
+            {% else %}
+                {{ include('@Swiftmailer/Collector/icon.svg') }}
```
Comment (stof) [inline Resources/views/Collector/swiftmailer.html.twig]: the icon does not exist. You forgot to include it in the PR.

File: Resources/views/Collector/swiftmailer.html.twig [hunk 3]
```diff
@@ -1,77 +1,175 @@
...
-                </div>
-                <div class="sf-toolbar-info-piece">
-                    <b>Is spooled ?</b>
-                    <span>{{ collector.isSpool(name) ? 'yes' : 'no' }}</span>
-                </div>
-            {% endfor %}
 
+            {% if profiler_markup_version == 1 %}
+                {% for name in collector.mailers %}
+                    <hr>
```
Comment (stof) [inline Resources/views/Collector/swiftmailer.html.twig]: should this ``hr`` be added even for the first element ?

File: Resources/views/Collector/swiftmailer.html.twig [hunk 4]
```diff
@@ -1,77 +1,175 @@
...
+        {% else %}
+            <div class="metrics">
+                {% for name in collector.mailers %}
+                    <div class="metric">
+                        <span class="value">{{ collector.messageCount(name) }}
+                        {% if collector.messageCount(name) > 0 %}
+                            <span class="unit">{{ collector.isSpool(name) ? 'spooled' : 'sent' }}</span>
+                        {% endif %}
+                        </span>
+                        <span class="label">sent messages</span>
```
Comment (stof) [inline Resources/views/Collector/swiftmailer.html.twig]: ``sent messages`` would look weird here. The block will say ``5 spooled sent messages`` or ``5 sent sent messages``

File: Resources/views/Collector/swiftmailer.html.twig [hunk 5]
```diff
@@ -1,77 +1,175 @@
...
+            <div class="metrics">
+                {% for name in collector.mailers %}
+                    <div class="metric">
+                        <span class="value">{{ collector.messageCount(name) }}
+                        {% if collector.messageCount(name) > 0 %}
+                            <span class="unit">{{ collector.isSpool(name) ? 'spooled' : 'sent' }}</span>
+                        {% endif %}
+                        </span>
+                        <span class="label">
+                            {{ name }} {{ collector.isDefaultMailer(name) ? ' (default app mailer)' : '' }}
```
Comment (stof) [inline Resources/views/Collector/swiftmailer.html.twig]: the ``: ''`` can be omitted in the ternary operator

File: Resources/views/Collector/swiftmailer.html.twig [hunk 6]
```diff
@@ -1,77 +1,175 @@
...
+            </div>
         {% else %}
-            <ul class="alt">
-                {% for message in collector.messages(name) %}
-                    <li class="{{ cycle(['odd', 'even'], loop.index) }}">
-                        {% for header in message.headers.all %}
-                            <pre>{{ header }}</pre>
-                        {% endfor %}
+            {% for message in collector.messages(name) %}
+                {% if collector.messages(name)|length > 1 %}
```
Comment (stof) [inline Resources/views/Collector/swiftmailer.html.twig]: use ``loop.length`` here. It will be more readable about which length is used IMO

File: Resources/views/Collector/swiftmailer.html.twig [hunk 7]
```diff
@@ -88,10 +189,10 @@
                                     {{- messagePart.body|e('html') }}
```
Comment (stof) [inline Resources/views/Collector/swiftmailer.html.twig]: explicit escaping could be removed here. We have auto-escaping.

File: Resources/views/Collector/swiftmailer.html.twig [hunk 8]
```diff
@@ -79,8 +177,11 @@
...
-                        {% for messagePart in message.children if messagePart.contentType == 'text/plain' or messagePart.contentType == 'text/html' %}
-                            <h4>Alternative part</h4>
+                    </div>
+
+                    {% for messagePart in message.children if messagePart.contentType == 'text/plain' or messagePart.contentType == 'text/html' %}
+                        <div class="card-block">
+                            <span class="label">Alternative part ({{ messagePart.contentType }})</span>
                             <pre>
                                 {%- if messagePart.charset %}
                                     {{- messagePart.body|e('html', messagePart.charset)|convert_encoding('UTF-8', messagePart.charset) }}
```
Comment (stof) [inline Resources/views/Collector/swiftmailer.html.twig]: I think we have a bug here: we are double-escaping it as ``convert_encoding`` is not marked as safe (which is logical) or as preserving safety.
The following would be better (relying on the auto-escaping for escaping):

```twig
{{- messagePart.body|convert_encoding('UTF-8', messagePart.charset) }}
```

The same apply a few lines above for the message body btw

**Comment by javiereguiluz** [comment]
> can you post a screenshot for a realistic use case for HTML parts of emails

Here it is a real example with a real HTML newsletter:

real-email

As you can see, the result is abominable. I propose to improve this as follows (in a separate pull request): I'd like to apply a `max-height` property on the email contents (as StackOverflow does for code blocks).

Another improvement (again in a separate PR) I'd like to do is to implement [this great idea](http://symfony.com/blog/new-in-symfony-2-8-redesigned-profiler#comment-20015) proposed by @garak to allow render the real HTML contents instead of just displaying the raw HTML tags.

File: Resources/views/Collector/swiftmailer.html.twig (hunk 4 above)
Comment (javiereguiluz) [inline Resources/views/Collector/swiftmailer.html.twig]: > sent messages would look weird here. The block will say 5 spooled sent messages or 5 sent sent messages

I agree with you. In https://github.com/javiereguiluz/SwiftmailerBundle/commit/7b5178decb95c0256e9c2ddbf14306c9b491bd30 I've redesigned this section a bit.

**Regular use case**: 1 mailer

Now we just say: `N spooled messages` or `N sent messages`

profiler-single-mailer

**Multiple-mailer user case**:

We display a table because we want to show a lot of details (mailer name, num. of messages, if they were sent or spooled, etc.)

profiler-multiple-mailers

profiler-multiple-mailers-messages

**Comment by javiereguiluz** [comment]
Status of this PR: I've implemented all the changes proposed by @stof, there is a new proposal pending approval (https://github.com/javiereguiluz/SwiftmailerBundle/commit/7b5178decb95c0256e9c2ddbf14306c9b491bd30) and there are two other ideas that will be implemented in a separate PR.

**Comment by stof** [comment]
Can you post a screenshot of your new proposal ? I don't have a project using this bundle to try it locally (and even less a 2.8 project)

File: Resources/views/Collector/swiftmailer.html.twig [hunk 9]
```diff
@@ -1,97 +1,182 @@
...
+                    <div class="sf-toolbar-info-piece">
+                        <b>{{ name }}</b>
+                        <span>{{ collector.messageCount(name) }}</span>
+                    </div>
+                    <div class="sf-toolbar-info-piece">
+                        <b>Is spooled?</b>
+                        <span>{{ collector.isSpool(name) ? 'yes' : 'no' }}</span>
+                    </div>
+
+                    {% if not loop.first %}
```
Comment (stof) [inline Resources/views/Collector/swiftmailer.html.twig]: are you sure about it ? Shouldn't it be either skipped for the last element instead, or displayed before the content for all except the first element ?

**Comment by javiereguiluz** [comment]
@stof the screenshots for the small redesign can be seen here: https://github.com/symfony/swiftmailer-bundle/pull/109#discussion_r40200606

**Comment by stof** [comment]
Except it is totally unreadable (the width of the image is 149px), so I cannot see how the top of the page looks like

**Comment by javiereguiluz** [comment]
@stof we are not understanding each other :)

The unreadable image of [this comment](https://github.com/symfony/swiftmailer-bundle/pull/109#issuecomment-142583358) shows the current design with a real email. It just shows how bad long emails are displayed. This problem will be solved later in another PR.

The small redesign that I propose for this PR is explained in the three screenshots included [in this other comment](https://github.com/symfony/swiftmailer-bundle/pull/109#discussion_r40200606).

**Comment by stof** [comment]
@javiereguiluz can you rebase your branch to fix conflicts ? the merge of #112 is causing one

**Comment by stof** [comment]
the small redesign looks good to me btw

**Comment by javiereguiluz** [comment]
As said above, the two pending issues will be solved in different PRs. I'm going to rebase this to be able to merge it ASAP. Thanks.

**Comment by javiereguiluz** [comment]
Rebased.

**Comment by fabpot** [comment]
Thank you @javiereguiluz.
